internal/rancher/management/v3/cluster.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright © 2023 - 2024 SUSE LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v3
+
+import (
+	"reflect"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	runtime "k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	// ClusterConditionAgentDeployed is the condition type for the agent deployed condition.
+	ClusterConditionAgentDeployed Cond = "AgentDeployed"
+	// ClusterConditionReady is the condition type for the ready condition.
+	ClusterConditionReady Cond = "Ready"
+)
+
+type (
+	// Cond represents a condition of a Rancher Cluster.
+	Cond string
+	// ClusterConditionType represents the type of a condition.
+	ClusterConditionType string
+)
+
+// Cluster is the struct representing a Rancher Cluster.
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+type Cluster struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ClusterSpec   `json:"spec"`
+	Status ClusterStatus `json:"status,omitempty"`
+}
+
+// ClusterSpec is the struct representing the specification of a Rancher Cluster.
+type ClusterSpec struct {
+	DisplayName        string `json:"displayName,omitempty"`
+	Description        string `json:"description,omitempty"`
+	FleetWorkspaceName string `json:"fleetWorkspaceName,omitempty"`
+}
+
+// ClusterStatus is the struct representing the status of a Rancher Cluster.
+type ClusterStatus struct {
+	Conditions []ClusterCondition `json:"conditions,omitempty"`
+}
+
+// ClusterList contains a list of ClusterList.
+// +kubebuilder:object:root=true
+type ClusterList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []Cluster `json:"items"`
+}
+
+// ClusterCondition is the struct representing a condition of a Rancher Cluster.
+type ClusterCondition struct {
```

Comment:
I think this custom logic is better to be replaced with `"sigs.k8s.io/cluster-api/util/conditions"` and implement the Getter interface: https://github.com/kubernetes-sigs/cluster-api/blob/96beb3577f5bbd5b762ac6e2cbfb05e9f8721aa9/util/conditions/getter.go#L29-L34

Reply:
I agree :+1:

Reply:
This custom logic is taken from `rancher/wrangler` [here](https://github.com/rancher/wrangler/blob/004384a454ec6062420e10439bf9fa136fa38e4d/pkg/condition/condition.go). Had to re-use the functions as they're not exported in the original code. I think it makes sense to stick to the way `rancher` does it, but I'm happy to change it if you think otherwise.

Reply:
The conditions is a kubernetes API, each project has to implement this equally to match API expectation. Wrangler is never a reliable source or example for anything, but here it does not matter what code to use, just importing this has 3 main benefits:
1. CAPI implementation is reliable, and is already used in our project.
2. If this copied code will have a security vulnerability, we will be responsible to fix it here. CAPI imported code will likely just be a dependency bump for us. 
3. Less boilerplate code in general opposed to copying.

Reply:
I think its probably best to follow conventions of upstream projects over wrangler (which we do not use).

Reply:
Thanks for the suggestion. Using the CAPI conditions is quite a neat solution.